File: placementprepration/tree/vertical order.cpp
#include<iostream>
#include <map>
#include <queue>
#include <vector>
#include <algorithm>
#include<set>
using namespace std;
struct Node{
    int data;
    Node*left;
    Node*right;
    Node(int val){
        this->data=val;
        this->left=NULL;
        this->right=NULL;
    }
};
Node*BuildTree(Node*root){
    int data;
    cout<<"Enter the value of data";
    cin>>data;
    if(data==-1){
        return NULL;
    }
    Node*root=new Node(data);
    cout<<"Enter the value of data in left of "<<data<<endl;
    root->left=BuildTree(root->left);
    cout<<"Enter the value of data in right of "<<data<<endl;
    root->right=BuildTree(root->right);
    return root;
}
vector<vector<int>>verticalTraversal(Node*root){
    map<int,map<int,multiset<int>>>node;
    queue<pair<Node*,pair<int,int>>>q;
    q.push({root,{0,0}});
    while(!q.empty()){
        auto p=q.front();
        q.pop();
        Node*node=p.first;
        int x=p.second.first,y=p.second.second;
        node[x][y].insert(node->data);
        if(node->left){
            q.push({node->left,{x-1,y+1}});
        }
        if(node->right){
            q.push({node->right,{x+1,y+1}});
        }
    }
    vector<vector<int>>ans;
    for(auto p:node){
        vector<int>col;
        for(auto q:p.second){
            for(auto it:q.second){
                col.push_back(it);
            }
        }
        ans.push_back(col);
    }
    return ans;

}

int main(){
    Node*root=NULL:
    root=BuildTree(root);
    

}

File: placementprepration/tree/vertical_order.py
from collections import deque, defaultdict


class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None


def vertical_order(root):
    if not root:
        return []
    col_map = defaultdict(list)
    q = deque([(root, 0)])
    while q:
        node, col = q.popleft()
        col_map[col].append(node.val)
        if node.left:
            q.append((node.left, col - 1))
        if node.right:
            q.append((node.right, col + 1))
    return [col_map[c] for c in sorted(col_map.keys())]


if __name__ == "__main__":
    r = Node(1)
    r.left = Node(2)
    r.right = Node(3)
    r.left.left = Node(4)
    r.left.right = Node(5)
    r.right.left = Node(6)
    r.right.right = Node(7)
    print(vertical_order(r))
